src/main/java/bio/terra/app/controller/Oauth2ApiController.java

```diff
@@ -0,0 +1,123 @@
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.client.RestTemplate;
+
+@Controller
+public class Oauth2ApiController {
+  static final String AUTHORIZE_ENDPOINT = "/oauth2/authorize";
+  static final String TOKEN_REFRESH_ENDPOINT = "/oauth2/token";
+  private static final String SCOPE_PARAM = "scope";
+  private static final String CLIENT_SECRET_PARAM = "client_secret";
+  private final RestTemplate restTemplate;
+  private final OpenIDConnectConfiguration openIDConnectConfiguration;
+
+  @Autowired
+  public Oauth2ApiController(OpenIDConnectConfiguration openIDConnectConfiguration) {
+    this.openIDConnectConfiguration = openIDConnectConfiguration;
+    restTemplate = new RestTemplate(new HttpComponentsClientHttpRequestFactory());
+  }
+
+  @RequestMapping(value = AUTHORIZE_ENDPOINT)
+  public String oauthAuthorize(HttpServletRequest request) {
+    return "redirect:"
+        + openIDConnectConfiguration.getAuthorizationEndpoint()
+        + "?"
+        + decorateAuthRedirectQueryParameters(request.getQueryString());
+  }
+
+  // Modify the query parameters that were passed to be compatible with configured Oauth flow
+  private String decorateAuthRedirectQueryParameters(String initialParams) {
+    List<NameValuePair> parameters = URLEncodedUtils.parse(initialParams, Charsets.UTF_8);
+
+    parameters.addAll(
+        URLEncodedUtils.parse(openIDConnectConfiguration.getExtraAuthParams(), Charsets.UTF_8));
+
+    parameters =
+        parameters.stream()
+            .map(
+                p -> {
+                  if (openIDConnectConfiguration.isAddClientIdToScope()
+                      && p.getName().equalsIgnoreCase(SCOPE_PARAM)) {
+                    return new BasicNameValuePair(
+                        p.getName(), p.getValue() + " " + openIDConnectConfiguration.getClientId());
+                  } else {
+                    return p;
+                  }
+                })
+            .collect(Collectors.toList());
+
+    return URLEncodedUtils.format(parameters, Charsets.UTF_8);
+  }
+
+  @RequestMapping(value = TOKEN_REFRESH_ENDPOINT, method = RequestMethod.POST)
+  public ResponseEntity<JsonNode> oauthTokenPost(HttpServletRequest request) {
+    String queryString =
+        StringUtils.isEmpty(request.getQueryString()) ? "" : "?" + request.getQueryString();
+    String actualEndpoint = openIDConnectConfiguration.getTokenEndpoint() + queryString;
+
+    String requestBody;
+    try (BufferedReader reader = request.getReader()) {
+      requestBody = addClientSecret(IOUtils.toString(reader));
+    } catch (IOException e) {
+      throw new BadRequestException("Could not process refresh request", e);
+    }
+
+    HttpHeaders headers = new HttpHeaders();
+    headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);
```

Comment:
Do we want to check that the _incoming_ request is `APPLICATION_FORM_URLENCODED` before sending it along to the token endpoint?

Reply:
whoa!  Great find!  Yeah, that's a create point.  I'll add that check

Reply:
what did the body look like?  I'm having trouble reproducing

Reply:
I just sent `{}`

Here's a repo, this kind of simulates what the proxy was doing (at least for Orch):
```
curl -v "https://oauth2.googleapis.com/token" -H "Content-Type: application/json" -d '{}&client_secret=foo'
```
I get a 400 with:
```
{
  "error": {
    "code": 400,
    "message": "Invalid JSON payload received. Parsing terminated before end of input.\n{}&client_secret=foo\n  ^",
    "status": "INVALID_ARGUMENT"
  }
}
```

Reply:
Mayybe this was an Orch-only bug because you are sending the Content-Type on the outgoing call (I was just proxying the header). Still good idea to validate though :)

Reply:
Hmmm...I can't seem to repro.  That being said, I'm happy to add a check that the body is form encoded